macrosynergy/panel/adjust_weights.py

```diff
@@ -0,0 +1,258 @@
+    )
+
+    norm_rows = df_weights_wide.sum(axis="columns").apply(lambda x: np.isclose(x, 1))
+    all_nan_rows = df_weights_wide.index[df_weights_wide.isnull().all(axis="columns")]
+
+    # assert that all rows sum to 1 or are all NaN
+    if not norm_rows.all() and all_nan_rows.size == 0:
+        raise Exception("Normalization failed; weights do not sum to 1")
+
+    return df_weights_wide
+
+
+def adjust_weights(
+    df: QuantamentalDataFrame,
+    weights: str,
+    adj_zns: str,
+    method: Callable[[Number], Number],
+    params: Dict[str, Any] = {},
+    cids: List[str] = None,
+    adj_name: str = "ADJWGT",
+):
+    """
+    Adjusts the weights of a given xcat by a given adjustment xcat using a given method.
+
+    Parameters
+    ----------
+    df : QuantamentalDataFrame
+        QuantamentalDataFrame with weights and adjustment xcats for all cids.
+    weights : str
+        Name of the xcat containing the weights.
+    adj_zns : str
+        Name of the xcat containing the adjustment factors.
+    method : Callable
+        Function that will be applied to the weights to adjust them. This function must
+        conform to `f(x: Number, *args) -> Number`.
+    params : Dict[str, Any], optional
+        Parameters to be passed to the method function. Default is {}.
+    cids : List[str], optional
+        List of cids to adjust. If None, all cids will be adjusted. Default is None.
+    adj_name : str, optional
+        Name of the resulting xcat. Default is "ADJWGT".
+    """
+
+    if not isinstance(df, QuantamentalDataFrame):
+        raise TypeError("df must be a QuantamentalDataFrame")
+
+    df = QuantamentalDataFrame(df)
+    result_as_categorical = df.InitializedAsCategorical
+
+    if cids is None:
+        cids = df["cid"].unique().tolist()
+
+    check_types(weights, adj_zns, method, params, cids)
+
+    df, r_xcats, r_cids = reduce_df(
+        df, cids=cids, xcats=[weights, adj_zns], intersect=True, out_all=True
+    )
+
+    check_missing_cids_xcats(weights, adj_zns, cids, r_xcats, r_cids)
+
+    df_weights_wide, df_adj_zns_wide = split_weights_adj_zns(df, weights, adj_zns)
+
+    df_weights_wide = normalize_weights(df_weights_wide)
+
+    dfw_result = adjust_weights_backend(
+        df_weights_wide, df_adj_zns_wide, method, params
+    )
+
+    all_nan_rows = dfw_result.index[dfw_result.isnull().all(axis="columns")]
+    if all_nan_rows.size > 0:
+        err = "The following dates have no data after applying the adjustment, and will be dropped:"
+        warnings.warn(f"{err} {all_nan_rows}")
+        dfw_result = dfw_result.dropna(how="all", axis="rows")
+
+    dfw_result = normalize_weights(dfw_result)
+
+    if dfw_result.isna().all().all():
+        raise ValueError(
+            "The resulting DataFrame is empty. Please check the input data,"
+            " the method function, and it's parameters."
+        )
+
+    dfw_result.columns = list(map(lambda x: f"{x}_{adj_name}", dfw_result.columns))
```

Comment:
To simplify
```python
dfw.result.columns += adj_name
```
instead of
```python
dfw_result.columns = list(map(lambda x: f"{x}_{adj_name}", dfw_result.columns))
```